core/primitives-core/src/account.rs

```diff
@@ -20,6 +23,173 @@ pub struct Account {
+    pub fn code_hash(&self) -> CryptoHash {
+        self.code_hash
+    }
+
+    #[cfg(feature = "protocol_feature_add_account_versions")]
+    pub fn code_hash(&self) -> CryptoHash {
+        match self {
+            Account::AccountV1(acc) => acc.code_hash,
+        }
+    }
+
+    #[cfg(not(feature = "protocol_feature_add_account_versions"))]
+    pub fn storage_usage(&self) -> StorageUsage {
+        self.storage_usage
+    }
+
+    #[cfg(feature = "protocol_feature_add_account_versions")]
+    pub fn storage_usage(&self) -> StorageUsage {
+        match self {
+            Account::AccountV1(acc) => acc.storage_usage,
+        }
+    }
+
+    #[cfg(not(feature = "protocol_feature_add_account_versions"))]
+    pub fn set_amount(&mut self, amount: Balance) {
+        self.amount = amount;
+    }
+
+    #[cfg(feature = "protocol_feature_add_account_versions")]
+    pub fn set_amount(&mut self, amount: Balance) {
+        match self {
+            Account::AccountV1(acc) => {
+                acc.amount = amount;
+            }
+        }
+    }
+
+    #[cfg(not(feature = "protocol_feature_add_account_versions"))]
+    pub fn set_locked(&mut self, locked: Balance) {
+        self.locked = locked;
+    }
+
+    #[cfg(feature = "protocol_feature_add_account_versions")]
+    pub fn set_locked(&mut self, locked: Balance) {
+        match self {
+            Account::AccountV1(acc) => {
+                acc.locked = locked;
+            }
+        }
+    }
+
+    #[cfg(not(feature = "protocol_feature_add_account_versions"))]
+    pub fn set_code_hash(&mut self, code_hash: CryptoHash) {
+        self.code_hash = code_hash;
+    }
+
+    #[cfg(feature = "protocol_feature_add_account_versions")]
+    pub fn set_code_hash(&mut self, code_hash: CryptoHash) {
+        match self {
+            Account::AccountV1(acc) => {
+                acc.code_hash = code_hash;
+            }
+        }
+    }
+
+    #[cfg(not(feature = "protocol_feature_add_account_versions"))]
+    pub fn set_storage_usage(&mut self, storage_usage: StorageUsage) {
+        self.storage_usage = storage_usage;
+    }
+
+    #[cfg(feature = "protocol_feature_add_account_versions")]
+    pub fn set_storage_usage(&mut self, storage_usage: StorageUsage) {
+        match self {
+            Account::AccountV1(acc) => {
+                acc.storage_usage = storage_usage;
+            }
+        }
+    }
+}
+
+#[cfg(feature = "protocol_feature_add_account_versions")]
+impl BorshDeserialize for Account {
+    fn deserialize(buf: &mut &[u8]) -> Result<Self, Error> {
+        if buf.len() == std::mem::size_of::<AccountV1>() {
+            // This should only ever happen if we have pre-transition account serialized in state
+            // See test_account_size
+            Ok(Account::AccountV1(<AccountV1 as BorshDeserialize>::deserialize(buf)?))
+        } else {
+            #[derive(BorshDeserialize)]
+            enum DeserializableAccount {
```

Comment:
this feels somewhat suboptimal to me. Essentially every time `Account` changes, this struct needs to be changed correspondingly. @matklad do you see a better way to do this?

Reply:
I actually discussed this with @matklad beforehand. Alternative is for Account to be an additional one-member tuple around enum, and I do think it is more hassle

Reply:
No, I don't see a better way to do this(doesn't mean there isn't one), and this generally seems OK to me. This seems to be inherent complexity: there are bytes in some specific format in the persistent storage, and there's a domain object which, due to backwards compatibility concerns, does not precisely match the "on-disk" format. Something somewhere needs to handle this discrepancy. serde's derive only really works out of the box if you don't have backwards comparability concerns. 

What does feel suboptimal to me is that *the rest* of the code needs to care about account versions now. Let me try to explain this. I am not sure if this is a problem or a feature, as, to know that, I need to understand the code in depth :-)

`Account` is a public enum with versions, so any code *could* care if account is `V1` or `V2`. Naively, I'd expect that we want to enforce that most of the code is oblivious to the version of an account. But, naturally, we must be backwards compatible with the data already existing in the world, parts of which were created before `V2` existed. The way I usually approach this is by making `Account` a *struct* which contains only the latest and greatest version of all the fields, and having a separate `AccountRepr` solely for purposes of deserializaion. 

To give a concrere example, let's say we had had `struct Account { name: String }` and then decided to evolve that to `struct Account { first_name:String, last_name: String }`. We could start with this:

```rust
pub struct Account {
    pub name: String,
}

impl Account {
    pub fn serialize(self) -> Vec<u8> {
        AccountRepr::from(self).serialize()
    }
    pub fn deserialize(data: Vec<u8>) -> Result<Account> {
        AccountRepr::deserialize(data).map(Account::from)
    }
}

// NB: not pub
#[derive(Serialize, Deserialize)]
enum AccountRepr {
    V1 { name: String },
}

impl From<AccountRepr> for Account {
    fn from(repr: AccountRepr) -> Self {
        match repr {
            AccountRepr::V1 { name } => Account { name },
        }
    }
}

impl From<Account> for AccountRepr {
    fn from(account: Account) -> Self {
        AccountRepr::V1 { name: account.name }
    }
}
```

and then, once we realize we neen first/last name, we change the code to

```rust
// V2

pub struct Account {
    pub first_name: String,
    pub last_name: String,
}

impl Account {
    pub fn serialize(self) -> Vec<u8> {
        AccountRepr::from(self).serialize()
    }
    pub fn deserialize(data: Vec<u8>) -> Result<Account> {
        AccountRepr::deserialize(data).map(Account::from)
    }
}

// NB: not pub
#[derive(Serialize, Deserialize)]
enum AccountRepr {
    V1 { name: String },
    V2 { first_name: String, last_name: String },
}

impl From<AccountRepr> for Account {
    fn from(repr: AccountRepr) -> Self {
        match repr {
            AccountRepr::V2 { first_name, last_name } => Account { first_name, last_name },
            AccountRepr::V1 { name } => {
                // The tricky bit: whats' the actual rule for staying backwards
                // compatible?
                //
                // Let's set the last name to an empty string, bug, eg,
                // splitting on `" "` might or might not be a valid ruler
                Account { first_name: name, last_name: String::new() }
            }
        }
    }
}

impl From<Account> for AccountRepr {
    fn from(account: Account) -> Self {
        AccountRepr::V2 { first_name: account.first_name, last_name: account.last_name }
    }
}
```

That way, there's single place in code that cares about existence of two different versions of accounts.


This scheme might, or might not work for our use-case. To work, there needs to be an injective mapping from old accounts to new ones. If there isn't one, and there's genuinely some business logic that needs to work differently for two acccounts types, we might need to add something like:

```rust
pub struct Account {
    pub first_name: String,
    pub last_name: String,
    pub creation_version: AccountCreationVersion
}

pub enum AccountCreationVersion {
    V1,
    V2
}
```

Making the `Account` itself an `enum` I think might be benefitial only if we explicitelly want all the logic to be different for different versions, so that each bit code is forced to `match` and than deal with v1 and v2 separatelly. 

Based on my priors of how the world usually works, I think there's a high probability that we are in case one (injective mapping) or case two (version flag is enough).

But, again, I know almost nothing about how the thing is actually used, so this might not be applicable.

Reply:
While I see merits of what you propose I think we need to be consistent as to which approach we use

Reply:
@matklad I don't think what you proposed works well for our use cases. For example, the change that @EgorKulikov does in this PR does not actually change the account structure (i.e, the new version has exactly the same structure) but rather changes the logic of how a field is updated. This means that while we can go with what you proposed, it does not save us much because we still need to do things differently based on protocol versions, which means that it is better to introduce getters and setters, as done in this PR.

Reply:
SGTM, def do not want to block on this. Though, I must the fact that the `enum` is public does make me uneasy. I'll take some time some time to read on the similar pattern in Block, to be more sure if this is essential or accidental complexity.